sdformat_urdf/src/sdformat_urdf.cpp

```diff
@@ -0,0 +1,629 @@
+  ignition::math::Pose3d link_pose;
+  if (!relative_joint_name.empty()) {
+    // urdf link pose is the location of the joint having it as a child
+    sdf::Errors pose_errors = sdf_link.SemanticPose().Resolve(link_pose, relative_joint_name);
+    if (!pose_errors.empty()) {
+      errors.insert(errors.end(), pose_errors.begin(), pose_errors.end());
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to get transform of link [" + sdf_link.Name() + "]");
+      return nullptr;
+    }
+  }
+
+  const ignition::math::Inertiald sdf_inertia = sdf_link.Inertial();
+  urdf_link->inertial = std::make_shared<urdf::Inertial>();
+  urdf_link->inertial->mass = sdf_inertia.MassMatrix().Mass();
+  // URDF doesn't have link pose concept, so add SDF link pose to inertial
+  urdf_link->inertial->origin = convert_pose(link_pose + sdf_inertia.Pose());
+  urdf_link->inertial->ixx = sdf_inertia.MassMatrix().Ixx();
+  urdf_link->inertial->ixy = sdf_inertia.MassMatrix().Ixy();
+  urdf_link->inertial->ixz = sdf_inertia.MassMatrix().Ixz();
+  urdf_link->inertial->iyy = sdf_inertia.MassMatrix().Iyy();
+  urdf_link->inertial->iyz = sdf_inertia.MassMatrix().Iyz();
+  urdf_link->inertial->izz = sdf_inertia.MassMatrix().Izz();
+
+  for (uint64_t vi = 0; vi < sdf_link.VisualCount(); ++vi) {
+    const sdf::Visual * sdf_visual = sdf_link.VisualByIndex(vi);
+    if (!sdf_visual) {
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to get visual on link [" + sdf_link.Name() + "]");
+      return nullptr;
+    }
+
+    auto urdf_visual = std::make_shared<urdf::Visual>();
+
+    urdf_visual->name = sdf_visual->Name();
+
+    // URDF visual is relative to link origin
+    ignition::math::Pose3d visual_pose;
+    sdf::Errors pose_errors = sdf_visual->SemanticPose().Resolve(visual_pose, sdf_link.Name());
+    if (!pose_errors.empty()) {
+      errors.insert(errors.end(), pose_errors.begin(), pose_errors.end());
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to get transform from visual [" + sdf_visual->Name() +
+        "] to link [" + sdf_link.Name() + "]");
+      return nullptr;
+    }
+    // URDF doesn't have link pose concept, so add SDF link pose to visual
+    urdf_visual->origin = convert_pose(link_pose + visual_pose);
+
+    urdf_visual->geometry = convert_geometry(*sdf_visual->Geom(), errors);
+    if (!urdf_visual->geometry) {
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to convert geometry on visual [" + sdf_visual->Name() + "]");
+      return nullptr;
+    }
+
+    const sdf::Material * sdf_material = sdf_visual->Material();
+    if (sdf_material) {
+      // TODO(sloretz) textures
+      // TODO(sloretz) error if any file names we can't resolve are given
+      auto urdf_material = std::make_shared<urdf::Material>();
+      // sdf materials don't have names, so assign it the visual's name and hope that's unique
```

Comment:
It seems to me that it's actually not that rare for a model to have two links each containing a visual with the same name (eg. `viz`) . How about if we concatenate the name of the link and the visual somehow? Then it should be unique since link names are unique.

Reply:
Concatenate link and visual names for material name in 197bbf8